azblob/request_common.go

```diff
@@ -0,0 +1,22 @@
+package azblob
+
+// ClientProvidedKeyOptions contains headers which may be be specified from service version 2019-02-02
+// or higher to encrypts the data on the service-side with the given key. Use of customer-provided keys
+// must be done over HTTPS. As the encryption key itself is provided in the request, a secure connection
+// must be established to transfer the key.
+// Note: Azure Storage does not store or manage customer provided encryption keys. Keys are securely discarded
+// as soon as possible after they’ve been used to encrypt or decrypt the blob data.
+type ClientProvidedKeyOptions struct {
+	EncryptionKey       *string                 // A Base64-encoded AES-256 encryption key value.
+	EncryptionKeySha256 *string                 // The Base64-encoded SHA256 of the encryption key.
+	EncryptionAlgorithm EncryptionAlgorithmType // Specifies the algorithm to use when encrypting data using the given key. Must be AES256.
+	EncryptionScope     *string
+}
+
+// InitClientProvidedKeyOptions function.
+// By default the value of encryption algorithm params is "AES256" for service version 2019-02-02 or higher.
+func InitClientProvidedKeyOptions(ek *string, eksha256 *string, es *string) (cpk *ClientProvidedKeyOptions) {
```

Comment:
>*ClientProvidedKeyOptions [](start = 81, length = 25)

we should return a literal struct instead of a pointer, right? Since that's what we take in for all the methods

Reply:
Thanks for pointing it out :)
I think there are couple of reasons why I did that but it is primarily due to optimizing for space. But then I decided to look at how memory allocation happens in Go and I concluded that stack allocation is cheap and heap allocation is expensive in Go. So try to reduce minimum number of pointers. Besides the size of struct is not large, so it makes sense to not use pointers. This behavior is slightly different in C++

Reference:
1. [Allocation efficiency in high-performance Go services](https://segment.com/blog/allocation-efficiency-in-high-performance-go-services/ )
2. [A visual guide to Go Memory Allocator from scratch](https://blog.learngoprogramming.com/a-visual-guide-to-golang-memory-allocator-from-ground-up-e132258453ed)